pageserver/src/tenant/layer_map.rs

```diff
@@ -272,50 +294,59 @@ where
     ///
     /// Helper function for BatchedUpdates::insert_historic
     ///
-    pub(self) fn insert_historic_noflush(&mut self, layer: Arc<L>) {
-        self.historic.insert(
-            historic_layer_coverage::LayerKey::from(&*layer),
-            Arc::clone(&layer),
-        );
-
+    pub(self) fn insert_historic_noflush(
+        &mut self,
+        layer: Arc<L>,
+        ctx: &RequestContext,
+    ) -> Result<()> {
+        let lr = layer.get_lsn_range();
+        for kr in layer.get_occupied_ranges(ctx)? {
+            self.historic.insert(
+                historic_layer_coverage::LayerKey {
+                    key: kr.start.to_i128()..kr.end.to_i128(),
+                    lsn: lr.start.0..lr.end.0,
+                    is_image: !layer.is_incremental(),
+                },
+                Arc::clone(&layer),
+            );
+        }
         if Self::is_l0(&layer) {
-            self.l0_delta_layers.push(layer);
+            self.l0_delta_layers.insert(LayerRef(layer.clone()));
         }
-
+        self.historic_layers.insert(LayerRef(layer));
         NUM_ONDISK_LAYERS.inc();
+        Ok(())
     }
 
     ///
     /// Remove an on-disk layer from the map.
     ///
     /// Helper function for BatchedUpdates::remove_historic
     ///
-    pub fn remove_historic_noflush(&mut self, layer: Arc<L>) {
-        self.historic
-            .remove(historic_layer_coverage::LayerKey::from(&*layer));
+    pub fn remove_historic_noflush(&mut self, layer: Arc<L>, ctx: &RequestContext) -> Result<()> {
+        let lr = layer.get_lsn_range();
+        for kr in layer.get_occupied_ranges(ctx)? {
+            self.historic.remove(historic_layer_coverage::LayerKey {
```

Comment:
We assume that no two layers have at least one identical occupied range. If they did, then removing one layer would also remove a part of the other layer.

The assumption holds today and I see no reason it would stop holding. You can even make a stronger statement that "if two layers have identical LSN range, they are disjoint". But IMO the assumption is worth documenting.

Reply:
> We assume that no two layers have at least one identical occupied range. 

I think you mean not range, but box - including LSN coordinates. Because definitely two layers may have the same key ranges  (i.e. all L0 layers).

> The assumption holds today and I see no reason it would stop holding. 

Actually I am not sure that this assumption is true. Originally the problems with overlapped layer was encountered more than year ago. I don't remember scenarios in which such layers can be generated, but @hlinnaka  has changed layer map to correctly handle this situation. So I believe that current contract is the following:if two layers overlap, then their overlaped segments contains the same content: it is not possible that some wal records corresponding to the particular page and LSN range are stored in one layer and some other wal records - in another layer. @hlinnaka , can you confirm it?

I just want to notice that adding information about holes (or their contrast - occupied ranges) doesn't change something in this assumptions: if we can have two layers with the same key/lsn range as well as we can have two layers with the same occupied key range. So if we need to more precisely document this assumption, it is not directly related with this PR.